src/ngraph/runtime/gpu/cudnn_emitter.cpp

```diff
@@ -343,3 +343,144 @@ size_t runtime::gpu::CUDNNEmitter::build_pooling(const GPURuntimeContext* ctx,
     m_primitive_emitter->cache(hash, primitive_index);
     return primitive_index;
 }
+
+size_t runtime::gpu::CUDNNEmitter::build_batchnorm(const runtime::gpu::GPURuntimeContext* ctx,
+                                                   const cudnnBatchNormMode_t& bn_op,
+                                                   const Prop& direction,
+                                                   const Shape& tensor_shape,
+                                                   const Shape& param_shape,
+                                                   double epsilon)
+{
+    // Assumes NC{d1...dN} format
+    std::stringstream ss;
+    ss.precision(std::numeric_limits<double>::digits10 + 2);
+
+    ss << "bn_op" << bn_op << "_dir" << static_cast<int>(direction) << "_ts"
+       << join(tensor_shape, "_") << "_ps" << join(param_shape, "_") << "_eps" << epsilon;
+    std::string hash = ss.str();
+    std::replace(hash.begin(), hash.end(), '.', '_');
+
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    if (epsilon < CUDNN_BN_MIN_EPSILON)
+    {
+        throw std::runtime_error("Batch Norm epsilon is less than CUDNN_BN_MIN_EPSILON");
+    }
+
+    cudnnTensorDescriptor_t derived_param_desc;
+    cudnnCreateTensorDescriptor(&derived_param_desc);
+    auto tensor_desc = runtime::gpu::cudnn_util::tensor_descriptor_from_shape(tensor_shape);
+    cudnnDeriveBNTensorDescriptor(derived_param_desc, tensor_desc, bn_op);
+
+    float alpha = 1.0, beta = 0.0;
+    gpu::primitive* batchnorm;
+    switch (direction)
+    {
+    case Prop::Inference:
+    {
+        batchnorm = new gpu::primitive{[=](void** inputs, void** outputs) {
```

Comment:
Not sure that you want [=] as that makes copies and then you take references of those copies. Not sure if that is even kosher. Use [&] instead
and read this https://www.cprogramming.com/c++11/c++11-lambda-closures.html

Reply:
I believe [=] here is actually quite important. This lambda is capturing by value because the references in this scope will have long since expired when this closure is executed (during graph execution). Capturing by reference would result in undefined behavior as the local variables, such as the tensor_shape, no longer exists on the stack. Thus it is better to copy by value. Additionally [=] only copies those items which are needed by the closure, it is essentially an easier way of doing a std::bind, so that this function call can have consistent signature `void(void**,void**)` with all of the other arguments capture by value at emission time.